### PR TITLE
Add concurrencykata module.

### PR DESCRIPTION
**Pull request by velkey1993**

### CODE REVIEW

File: concurrencykata/src/main/java/kata/concurrency/testframework/api/Test.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+public class Test {
+
+    private static final OkHttpClient OK_HTTP_CLIENT = new OkHttpClient();
+    private static final ThreadLocal<TestContext> TEST_CONTEXT_THREAD_LOCAL = new ThreadLocal<>();
+
+    protected static void init() {
+        TEST_CONTEXT_THREAD_LOCAL.set(new TestContext());
+    }
+
+    protected static void destroy() {
```
Comment (kpjanos) [inline concurrencykata/src/main/java/kata/concurrency/testframework/api/Test.java]: This method is not part of the original API. Not an issue though.

File: concurrencykata/src/main/java/kata/concurrency/testframework/api/Test.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        checkTestState(TestStatus.GIVEN);
+        TEST_CONTEXT_THREAD_LOCAL.get().getGivenContext().setPath(path);
+    }
+
+    public static void setBaseUrl(String url) {
+        checkTestState(TestStatus.GIVEN);
+        TEST_CONTEXT_THREAD_LOCAL.get().getGivenContext().setBaseUrl(url);
+    }
+
+    public static <T> void setResponseDeserializer(Function<String, T> deserializer) throws IOException {
```
Comment (kpjanos) [inline concurrencykata/src/main/java/kata/concurrency/testframework/api/Test.java]: [mf] 'set' method invokes HTTP call, furthermore it executes the deserialization. This really violates single responsibility.

File: concurrencykata/src/main/java/kata/concurrency/testframework/api/Test.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+    public static <T> void assertThat(Consumer<T> assertion) {
+        checkTestState(TestStatus.WHEN, TestStatus.THEN);
+        TestContext testContext = TEST_CONTEXT_THREAD_LOCAL.get();
+        T response = (T) testContext.getWhenContext().getResponse();
+        assertion.accept(response);
+        testContext.setTestStatus(TestStatus.THEN);
+    }
+
+    private static void checkTestState(TestStatus... expectedTestStatuses) {
```
Comment (kpjanos) [inline concurrencykata/src/main/java/kata/concurrency/testframework/api/Test.java]: [optional] This is bit confusing that this method accepts multiple TestStatus-es, can be done with only passing the test step status and checking that violates the current status. Also combining in the setTestStatus would result in a less error prone code.

File: concurrencykata/src/main/java/kata/concurrency/testframework/api/Test.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        TestStatus actualTestStatus = TEST_CONTEXT_THREAD_LOCAL.get().getTestStatus();
+        if (!Arrays.asList(expectedTestStatuses).contains(actualTestStatus)) {
+            throw new WrongTestOrderException();
+        }
+    }
+
+    private static URL buildURL(TestGivenContext testGivenContext) {
```
Comment (kpjanos) [inline concurrencykata/src/main/java/kata/concurrency/testframework/api/Test.java]: [mf] This seems like a method which should be in TestGivenContext.

File: concurrencykata/src/main/java/kata/concurrency/testframework/api/TestRunner.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+    private static final Logger LOGGER = Logger.getLogger(TestRunner.class.getName());
+
+    private final int maxThreadCount;
+
+    public TestRunner(int maxThreadCount) {
+        this.maxThreadCount = maxThreadCount;
+    }
+
+    public void run(Class<?>... testClasses) throws IllegalAccessException, InstantiationException,
```
Comment (kpjanos) [inline concurrencykata/src/main/java/kata/concurrency/testframework/api/TestRunner.java]: [mf] violates single responsibility. This method&class does way too much.
- searching test methods.
- scheduling of test case runs
- execute test cases
- result reporting

File: concurrencykata/src/main/java/kata/concurrency/testframework/api/TestRunner.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+        List<Object> testClassInstances = getTestClassInstances(testClasses).stream()
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+
+        List<InvocationContext> invocationContexts = testClassInstances.stream()
+                .flatMap(this::getInvocationContextStream)
+                .collect(Collectors.toList());
+
+        List<Future<TestResult>> futures = filterContextByAnnotation(invocationContexts, TestCase.class).stream()
```
Comment (kpjanos) [inline concurrencykata/src/main/java/kata/concurrency/testframework/api/TestRunner.java]: [question] is there any reason for not chaining the stream together? stream()...(toList) repeatedly.

File: concurrencykata/src/main/java/kata/concurrency/testframework/api/TestRunner.java [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    private List<Object> getTestClassInstances(Class<?>[] testClasses) throws InstantiationException, IllegalAccessException {
+        List<Object> instances = new ArrayList<>();
+        for (Class<?> testClass : testClasses) {
+            instances.add(testClass.newInstance());
+        }
+        return instances;
+    }
+
+    private Stream<InvocationContext> getInvocationContextStream(Object instance) {
+        return Arrays.stream(instance.getClass().getDeclaredMethods())
```
Comment (kpjanos) [inline concurrencykata/src/main/java/kata/concurrency/testframework/api/TestRunner.java]: [optional] how about https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/reflect/MethodUtils.html#getMethodsListWithAnnotation-java.lang.Class-java.lang.Class- ?

File: concurrencykata/src/main/java/kata/concurrency/testframework/api/TestRunner.java [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        }
+    }
+
+    private List<InvocationContext> filterContextByAnnotation(List<InvocationContext> invocationContexts, Class<? extends Annotation> annotation) {
+        return invocationContexts.stream()
+                .filter(context -> context.getMethod().isAnnotationPresent(annotation))
+                .collect(Collectors.toList());
+    }
+
+    private static class TestRunnable implements Callable<TestResult> {
```
Comment (kpjanos) [inline concurrencykata/src/main/java/kata/concurrency/testframework/api/TestRunner.java]: [mf] naming: a runnable which is a callable...

File: concurrencykata/src/main/java/kata/concurrency/testframework/internal/model/TestContext.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
+package kata.concurrency.testframework.internal.model;
+
+public class TestContext {
+
+    private final TestGivenContext givenContext;
+    private final TestWhenContext whenContext;
```
Comment (kpjanos) [inline concurrencykata/src/main/java/kata/concurrency/testframework/internal/model/TestContext.java]: [question] why dont we have TestThenContext? would be nice matching of the domain

File: concurrencykata/src/main/java/kata/concurrency/testframework/api/TestRunner.java (hunk 6 above)
Comment (velkey1993) [reply concurrencykata/src/main/java/kata/concurrency/testframework/api/TestRunner.java]: [answer] as far as I remember correctly the only reason for this was that _invocationContexts_ was being used by the _preDestroyTestClasses_ method.

File: concurrencykata/src/main/java/kata/concurrency/testframework/internal/model/TestContext.java (hunk 9 above)
Comment (velkey1993) [reply concurrencykata/src/main/java/kata/concurrency/testframework/internal/model/TestContext.java]: [answer] at first i've created this class but based on the superfluity it has been deleted.

**Pull request closed without merging**